File: packages/game/src/systems/AbilitySystem.ts
import type { ECSEntity } from '@/state/ecs';
import abilitiesData from '../data/abilities.json';

export interface Ability {
  /** Unique ability ID */
  id: string;
  /** Display name */
  name: string;
  /** Description of the ability */
  description: string;
  /** Cost to use (energy, mana, etc.) */
  cost: number;
  /** Cooldown duration in milliseconds */
  cooldown: number;
  /** Effect type */
  effectType: 'damage' | 'buff' | 'debuff' | 'heal' | 'utility';
  /** Effect value (damage amount, heal amount, etc.) */
  effectValue: number;
  /** Additional properties */
  properties?: Record<string, unknown>;
}

export interface AbilityCooldownState {
  /** Ability ID */
  abilityId: string;
  /** Timestamp when cooldown ends */
  endsAt: number;
}

export interface AbilityExecutionResult {
  /** Whether the ability was successfully executed */
  success: boolean;
  /** Reason for failure (if any) */
  failureReason?: string;
  /** Effect applied */
  effect?: {
    type: Ability['effectType'];
    value: number;
    target: ECSEntity;
  };
}

// Module-scoped ability database
const abilityDatabase: Record<string, Ability[]> = abilitiesData as Record<string, Ability[]>;

/**
 * Get abilities available to a character.
 *
 * @param characterId - The character ID (e.g. 'kai', 'vera')
 * @returns Array of abilities
 */
export function getCharacterAbilities(characterId: string): Ability[] {
  return abilityDatabase[characterId] || [];
}

/**
 * Check if an ability is currently on cooldown.
 *
 * @param cooldownState - The cooldown state for the ability
 * @returns True if on cooldown
 */
export function isOnCooldown(cooldownState?: AbilityCooldownState): boolean {
  if (!cooldownState) return false;
  return Date.now() < cooldownState.endsAt;
}

/**
 * Apply cooldown to an ability.
 *
 * @param ability - The ability used
 * @returns New cooldown state
 */
export function applyCooldown(ability: Ability): AbilityCooldownState {
  return {
    abilityId: ability.id,
    endsAt: Date.now() + ability.cooldown,
  };
}

/**
 * Execute an ability.
 *
 * @param caster - The entity casting the ability
 * @param target - The target entity
 * @param ability - The ability to execute
 * @param cooldowns - Current cooldowns for the caster
 * @returns Result of execution
 */
export function executeAbility(
  caster: ECSEntity,
  target: ECSEntity,
  ability: Ability,
  cooldowns: AbilityCooldownState[]
): AbilityExecutionResult {
  // Check if ability is on cooldown
  const cooldownState = cooldowns.find((cd) => cd.abilityId === ability.id);
  if (isOnCooldown(cooldownState)) {
    return {
      success: false,
      failureReason: 'Ability is on cooldown',
    };
  }

  // Check resource cost
  if (ability.cost > 0) {
    if ((caster.mana ?? 0) < ability.cost) {
      return {
        success: false,
        failureReason: 'Insufficient mana',
      };
    }
    // Deduct cost
    caster.mana = (caster.mana ?? 0) - ability.cost;
  }

  // Apply ability effect
  let effect: { type: Ability['effectType']; value: number; target: ECSEntity } | undefined;

  switch (ability.effectType) {
    case 'damage':
      if (target.health !== undefined) {
        target.health = Math.max(0, target.health - ability.effectValue);
        effect = { type: 'damage', value: ability.effectValue, target };
      }
      break;

    case 'heal':
      if (target.health !== undefined && target.stats?.structure !== undefined) {
        const maxHealth = target.stats.structure;
        target.health = Math.min(maxHealth, target.health + ability.effectValue);
        effect = { type: 'heal', value: ability.effectValue, target };
      }
      break;

    case 'buff':
    case 'debuff':
    case 'utility':
      // These would be implemented with a buff/debuff system
      // For now, just mark as successful and return the effect for visualization
      effect = { type: ability.effectType, value: ability.effectValue, target };
      break;
  }

  if (!effect) {
    return {
      success: false,
      failureReason: 'Invalid target for ability',
    };
  }

  return {
    success: true,
    effect,
  };
}

File: Assets/Scripts/Systems/Abilities/AbilitySystem.cs
using Unity.Burst;
using Unity.Collections;
using Unity.Entities;
using Unity.Mathematics;
using NeoTokyo.Components.Abilities;
using NeoTokyo.Components.Stats;
using static Unity.Mathematics.math;

namespace NeoTokyo.Systems.Abilities
{
    /// <summary>
    /// System that processes ability cooldowns
    /// </summary>
    [BurstCompile]
    [UpdateInGroup(typeof(SimulationSystemGroup))]
    public partial struct AbilityCooldownSystem : ISystem
    {
        [BurstCompile]
        public void OnUpdate(ref SystemState state)
        {
            float deltaTime = SystemAPI.Time.DeltaTime;

            new UpdateCooldownsJob
            {
                DeltaTime = deltaTime
            }.ScheduleParallel();
        }

        [BurstCompile]
        partial struct UpdateCooldownsJob : IJobEntity
        {
            public float DeltaTime;

            void Execute(ref AbilityCooldown cooldown)
            {
                if (cooldown.RemainingTime > 0f)
                {
                    cooldown.RemainingTime = math.max(0f, cooldown.RemainingTime - DeltaTime);
                }
            }
        }
    }

    /// <summary>
    /// System that processes ability execution requests
    /// </summary>
    [BurstCompile]
    [UpdateInGroup(typeof(SimulationSystemGroup))]
    [UpdateAfter(typeof(AbilityCooldownSystem))]
    public partial struct AbilityExecutionSystem : ISystem
    {
        public void OnCreate(ref SystemState state)
        {
            state.RequireForUpdate<AbilityExecuteRequest>();
            state.RequireForUpdate<EndSimulationEntityCommandBufferSystem.Singleton>();
        }

        public void OnUpdate(ref SystemState state)
        {
            var ecb = SystemAPI.GetSingleton<EndSimulationEntityCommandBufferSystem.Singleton>()
                .CreateCommandBuffer(state.WorldUnmanaged);

            foreach (var (request, entity) in
                SystemAPI.Query<RefRO<AbilityExecuteRequest>>()
                    .WithEntityAccess())
            {
                ProcessAbilityRequest(ref state, request.ValueRO, entity, ref ecb);
            }
        }

        private void ProcessAbilityRequest(
            ref SystemState state,
            AbilityExecuteRequest request,
            Entity requestEntity,
            ref EntityCommandBuffer ecb)
        {
            // Get ability data
            if (!SystemAPI.HasComponent<AbilityData>(request.Ability))
            {
                ecb.DestroyEntity(requestEntity);
                return;
            }

            var abilityData = SystemAPI.GetComponent<AbilityData>(request.Ability);

            // Check cooldown
            if (SystemAPI.HasComponent<AbilityCooldown>(request.Ability))
            {
                var cooldown = SystemAPI.GetComponent<AbilityCooldown>(request.Ability);
                if (cooldown.IsOnCooldown)
                {
                    CreateFailureResult(ref ecb, request, "Ability is on cooldown");
                    ecb.DestroyEntity(requestEntity);
                    return;
                }
            }

            // Check resource cost
            if (SystemAPI.HasComponent<ResourcePool>(request.Caster))
            {
                var resources = SystemAPI.GetComponent<ResourcePool>(request.Caster);
                if (!resources.CanAfford(abilityData.Cost))
                {
                    CreateFailureResult(ref ecb, request, "Insufficient mana");
                    ecb.DestroyEntity(requestEntity);
                    return;
                }

                // Deduct cost
                resources.Spend(abilityData.Cost);
                ecb.SetComponent(request.Caster, resources);
            }

            // Apply cooldown
            if (SystemAPI.HasComponent<AbilityCooldown>(request.Ability))
            {
                var cooldown = new AbilityCooldown
                {
                    RemainingTime = abilityData.CooldownDuration,
                    TotalDuration = abilityData.CooldownDuration
                };
                ecb.SetComponent(request.Ability, cooldown);
            }

            // Apply effect based on type
            bool success = ApplyAbilityEffect(ref state, abilityData, request.Target, ref ecb);

            if (success)
            {
                CreateSuccessResult(ref ecb, request, abilityData);
            }
            else
            {
                CreateFailureResult(ref ecb, request, "Invalid target");
            }

            ecb.DestroyEntity(requestEntity);
        }

        private bool ApplyAbilityEffect(
            ref SystemState state,
            AbilityData ability,
            Entity target,
            ref EntityCommandBuffer ecb)
        {
            switch (ability.EffectType)
            {
                case AbilityEffectType.Damage:
                    // Damage reduces Health.Current (not Structure which is base defense/HP capacity)
                    if (SystemAPI.HasComponent<Health>(target))
                    {
                        var health = SystemAPI.GetComponent<Health>(target);
                        health.Current = math.max(0, health.Current - ability.EffectValue);
                        ecb.SetComponent(target, health);
                        return true;
                    }
                    break;

                case AbilityEffectType.Heal:
                    // Heal restores Health.Current up to max
                    if (SystemAPI.HasComponent<Health>(target))
                    {
                        var health = SystemAPI.GetComponent<Health>(target);
                        health.Current = math.min(health.Max, health.Current + ability.EffectValue);
                        ecb.SetComponent(target, health);
                        return true;
                    }
                    break;

                case AbilityEffectType.Buff:
                case AbilityEffectType.Debuff:
                case AbilityEffectType.Utility:
                    // These would trigger buff/debuff system
                    return true;
            }

            return false;
        }

        private void CreateSuccessResult(
            ref EntityCommandBuffer ecb,
            AbilityExecuteRequest request,
            AbilityData ability)
        {
            var resultEntity = ecb.CreateEntity();
            ecb.AddComponent(resultEntity, new AbilityExecuteResult
            {
                Success = true,
                EffectType = ability.EffectType,
                EffectValue = ability.EffectValue,
                Target = request.Target,
                FailureReason = default
            });
        }

        private void CreateFailureResult(
            ref EntityCommandBuffer ecb,
            AbilityExecuteRequest request,
            FixedString64Bytes reason)
        {
            var resultEntity = ecb.CreateEntity();
            ecb.AddComponent(resultEntity, new AbilityExecuteResult
            {
                Success = false,
                EffectType = default,
                EffectValue = 0,
                Target = request.Target,
                FailureReason = reason
            });
        }
    }

    /// <summary>
    /// System that handles resource regeneration
    /// </summary>
    [BurstCompile]
    [UpdateInGroup(typeof(SimulationSystemGroup))]
    public partial struct ResourceRegenSystem : ISystem
    {
        [BurstCompile]
        public void OnUpdate(ref SystemState state)
        {
            float deltaTime = SystemAPI.Time.DeltaTime;

            new RegenResourcesJob
            {
                DeltaTime = deltaTime
            }.ScheduleParallel();
        }

        [BurstCompile]
        partial struct RegenResourcesJob : IJobEntity
        {
            public float DeltaTime;

            void Execute(ref ResourcePool resources)
            {
                if (resources.Current < resources.Maximum && resources.RegenRate > 0f)
                {
                    int regenAmount = (int)(resources.RegenRate * DeltaTime);
                    resources.Current = math.min(resources.Maximum, resources.Current + regenAmount);
                }
            }
        }
    }
}
